framework/src/media/media_player.c

```diff
@@ -0,0 +1,328 @@
+		PLAY_UNLOCK();
+		return PLAY_ERROR_INVALID_STATE;
+	}
+
+	char instruction = ST_PROTO_STOP;
+	send(g_pc.fd, &instruction, 1, 0);
+
+	g_pc.state = PLAY_STOPPING;
+
+	PLAY_UNLOCK();
+
+	return MEDIA_OK;
+}
+
+play_result_t media_pause_play(void)
+{
+	PLAY_LOCK();
+
+	if (g_pc.state != PLAY_RUNNING) {
+		PLAY_UNLOCK();
+		return MEDIA_ERROR_INVALID_PARAMETER;
+	}
+
+	char instruction = ST_PROTO_PAUSE;
+	send(g_pc.fd, &instruction, 1, 0);
+
+	g_pc.state = PLAY_PAUSED;
+
+	PLAY_UNLOCK();
+
+	return MEDIA_OK;
+}
+
+play_result_t media_resume_play(void)
+{
+	if (g_pc.state != PLAY_PAUSED) {
+		return MEDIA_ERROR_INVALID_PARAMETER;
+	}
+
+	char instruction = ST_PROTO_RESUME;
+	send(g_pc.fd, &instruction, 1, 0);
+
+	g_pc.state = PLAY_RUNNING;
+	sem_post(&g_pc.sem_pause);
+
+	return MEDIA_OK;
+}
+
+play_result_t media_set_vol(unsigned char vol)
+{
+	return vol;
+}
+
+unsigned char media_get_vol(void)
+{
+	return 0;
+}
+
+play_result_t media_set_position(int time_ms)
+{
+	return MEDIA_OK;
+}
+
+int media_get_position(void)
+{
+	return 0;
+}
+
+bool media_is_available(void)
+{
+	return true;
+}
+
+bool media_is_playing(void)
+{
+	return (g_pc.state == PLAY_RUNNING);
+}
+
+/**
+ * Internal Functions
+ */
+
+int read_data_pcm_write(char *buffer)
+{
+	int readed = 0;
+	int remain;
+	int ret;
+
+	while (readed < g_pc.buffer_size) {
+		if (IS_SOCKET(g_pc.fd)) {
+			ret = recv(g_pc.fd, buffer + readed, g_pc.buffer_size - readed, 0);
+		} else {
+			ret = read(g_pc.fd, buffer + readed, g_pc.buffer_size - readed);
+		}
+		readed += ret;
+		if (ret == 0) {
+			break;
+		}
+	}
+	g_pc.download_size += readed;
+	if (readed) {
+		remain = pcm_bytes_to_frames(g_pc.pcmout, readed);
+		while (remain > 0) {
+			ret = pcm_writei(g_pc.pcmout, buffer + readed - pcm_frames_to_bytes(g_pc.pcmout, remain), remain);
+			if (ret > 0) {
+				remain -= ret;
+			} else {
+				return ret;
+			}
+		}
+	}
+
+	return readed;
+}
+
+int media_playing(void *args)
+{
+	int ret;
+	char *buffer = NULL;
+	buffer = (char *)malloc(g_pc.buffer_size);
+
+	if (!buffer) {
+		PLAY_LOCK();
+		g_pc.state = PLAY_AVAILABLE;
+		PLAY_UNLOCK();
+		return PLAY_ERROR_MALLOC;
+	}
+
+	while (g_pc.thread_running) {
+		PLAY_LOCK();
+		switch (g_pc.state) {
+		case PLAY_RUNNING:
+			ret = read_data_pcm_write(buffer);
+			if (ret == 0) {
+				g_pc.state = PLAY_DRAINING;
+			} else if (ret < 0) {
+				g_pc.state = PLAY_STOPPING;
+			}
+			break;
+
+		case PLAY_PAUSED:
+			PLAY_UNLOCK();
+			sem_wait(&g_pc.sem_pause);
+			PLAY_LOCK();
+			break;
+
+		case PLAY_STOPPING:
+			g_pc.thread_running = false;
+			pthread_join(g_pc.pth, NULL);
+			break;
+
+		case PLAY_DRAINING:
+			pcm_drain(g_pc.pcmout);
+			g_pc.state = PLAY_STOPPING;
+
+		default:
+			g_pc.thread_running = false;
+			g_pc.state = PLAY_STOPPING;
+			pthread_join(g_pc.pth, NULL);
+			break;
+		}
+		PLAY_UNLOCK();
+
+		// For context-switch
+		sched_yield();
+	}
+
+	free(buffer);
+
+	PLAY_LOCK();
+	g_pc.state = PLAY_AVAILABLE;
+	PLAY_UNLOCK();
+

```

Comment:
pcm object once opened , is never closed. This could lead to many problems , if some other application is trying to use audio device via tinyalsa and not via mediaplayer, then it would always return an error saying pcm device not available. This is like leaking pcm device forever.

Reply:
Added pcm_close at the bottom of thread function.